pandas/core/indexes/base.py

```diff
@@ -2246,6 +2246,25 @@ def _get_reconciled_name_object(self, other):
             return self._shallow_copy(name=name)
         return self
 
+    def _union_incompatible_dtypes(self, other):
+        """
+        Casts this and other index to object dtype to allow the formation
+        of a union between incompatible types.
+        """
+        this = self.astype(object)
+        # call Index for when `other` is list-like
+        other = Index(other).astype(object)
```

Comment:
I think better to do
``other = Index(other, dtype=object, copy=False)``

Reply:
or maybe just ``ensure_index``

Reply:
with `ensure_index` would have to be:

    ensure_index([1,2,3,4]).astype(object)

So perhaps the first suggestion would be better?

Reply:
yeah try the first one it would only make 1 copy (and not 2)

Reply:
so it turns out `other = Index(other, dtype=object...)` is not quite the same as `other = Index(other).astype(object)` when other is not an Index object.

Take this case for example:

    >>> other = [numpy.datetime64('2000-01-03T00:00:00.000000000'), numpy.datetime64('2000-01-04T00:00:00.000000000')]
    >>> pd.Index(other).astype(object)[0]
    Timestamp('2000-01-03 00:00:00')
    >>> pd.Index(other, dtype=object)[0]
    numpy.datetime64('2000-01-03T00:00:00.000000000')

I can see merits in both ways of doing it. The first allows pandas to cast to the proper pandas compatible type before unioning as object, while the latter is True to the original dtype.

I personally like the first since I can't imagine any user would ever not want the appropriate pandas compatible types.